patch for xrOS

johnzhou721 opened the pull request:
Patch for xrOS in Git commit format as mentioned in https://github.com/beeware/Python-Apple-support/pull/269.

**EDIT** -- new PR for PAS is at https://github.com/beeware/Python-Apple-support/pull/270

Context for more visitors: PAS = beeware/Python-Apple-Support, xrOS is internal codename in Apple for visionOS. This patch adds support to an already-patched version of CPython to support for visionOS.

johnzhou721 committed:
patch for xrOS

freakboy3742 commented on configure.ac:
```diff
@@ -480,8 +492,11 @@ if test -z "$CXX"; then
 
 		aarch64-apple-watchos*-simulator) CXX=arm64-apple-watchos-simulator-clang++ ;;
-		aarch64-apple-watchos*)           CXX=arm64_32-apple-watchos-clang++ ;;
+		aarch64-apple-watchos*)           CXX=arm64-apple-watchos-clang++ ;;
```
This one was correct.
```suggestion
		aarch64-apple-watchos*)           CXX=arm64_32-apple-watchos-clang++ ;;
```

johnzhou721 replied:
Oops. Thought I was correcting a typo for visionOS

freakboy3742 reviewed (changes requested):
The broad strokes of this PR look right; but there's a few details to sort out.

* There's definitely an aclocal version issue; there shouldn't be a need for any changes to `aclocal.m4`. If you use the `Tools/build/regen-configure.sh` script mentioned at the top of `configure.ac`, that will guarantee all the right tool versions are in use (using Docker to provide a clean development environment).

* The `xros` vs `visionOS` naming convention flagged in your other visionOS PR. There are definitely some places here where `xros` will need to be used (most notably, as `XROS_DEPLOYMENT_TARGET` - but `ac_sys_system` should definitely be resolving to `visionOS` so that `sys.platform` returns `visionos`; and the modifications in `platform` should also be using `visionOS` naming.

* There are some consistency issues with flagging the minimum visionOS version as 4.0, and the question of how that version is to be passed into the compiler - the compiler shims may need to be modified to include `XROS_DEPLOYMENT_TARGET` as part of the `-target` definition.

* There's no need to provide `arm64_32` or `x86_64` variants of the compiler shims, as those platforms don't exist for visionOS - and there's no need for the `pyconfig.h` shim either, as there's only 1 architecture for vision OS.

johnzhou721 committed:
Update configure.ac

Co-authored-by: Russell Keith-Magee <russell@keith-magee.com>

johnzhou721 commented:
Again, sorry for 4.0; I should've had more diligence when copy + pasting

johnzhou721 commented:
In addition -- I have some stuff I want to clarify --- I copy-pasted a lot of stuff from watchOS -- there's some stuff unavailable only for TV and Watch -- in that case, I think we should test those tv- and watchOS only stuff on visionOS and see if they're _actually_ unavailable.

johnzhou721 committed:
modify to visionOS

johnzhou721 committed:
Merge remote-tracking branch 'origin/3.14-patched' into 3.14-patched

johnzhou721 committed:
autoreconf using the docker i guess

johnzhou721 committed:
file business

johnzhou721 committed:
config.sub

johnzhou721 committed:
filenames

johnzhou721 committed:
some changes

johnzhou721 commented:
OK I have configure running locally at this point, I changed some more stuff that I need to push

johnzhou721 commented:
visionos deployment target -- that variable i think is fine

Besides the SDK everything else could use visionOS; I kept the xros target triples just in case someone needs to compile them using xros

johnzhou721 committed:
Configure and make now runs correctly

johnzhou721 committed:
remove info.plist i guess

johnzhou721 commented:
OK somehow make clean does not remove the generated info.plist. Haven't tried make install yet but configure and make could run (idk if correctly)

johnzhou721 committed:
update visionos in lib

johnzhou721 commented:
Not very sure about the latest commit-- no way to test at this moment -- I will need to generate a patch and feed it into py-apple-support

johnzhou721 commented:
@freakboy3742  Ping

freakboy3742 commented:
> @freakboy3742 Ping

As I mentioned yesterday - I get notified on every change to every BeeWare repo; there's no need to ping me directly. 

Also - timezones are a thing. I've literally just woken up :-)

freakboy3742 commented on .gitignore:
```diff
@@ -84,6 +84,8 @@ tvOS/Frameworks
 watchOS/Frameworks
 watchOS/Resources/Info.plist
+tvOS/Frameworks
+tvOS/Resources/Info.plist
```
I'm assuming this is a typo?

johnzhou721 replied:
Yep

freakboy3742 commented on Lib/sysconfig/__init__.py:
```diff
@@ -727,6 +727,10 @@ def get_platform():
             osname = sys.platform
             machine = sys.implementation._multiarch
+        elif sys.platform == "visionos":
+            release = get_config_vars().get("VISIONOS_DEPLOYMENT_TARGET", "4.0")
```
The 4.0 value can't be correct.

Also - we need to be consistent on whether the option is `XROS_DEPLOYMENT_TARGET` or `VISIONOS_DEPLOYMENT_TARGET`. All documentation I can find says that the option that Apple honors is `XROS_...` so we should stick with that one.

johnzhou721 replied:
Hi! Can you give an example of what Apple tooling accepts this? I am currently on the go and cannot work on this

freakboy3742 replied:
I haven't found public documentation of this setting, but it's declared in`SDKSettings.json` in the root of the visionOS SDK (and the analogous value in iPhone SDKs is `IPHONEOS_DEPLOYMENT_TARGET`)

johnzhou721 replied:
Done!

(base) johnzhou@Johns-MBP cpython-fb37 % grep -r --exclude-dir=.git "VISIONOS_DEPLOYMENT_TARGET" .

(base) johnzhou@Johns-MBP cpython-fb37 %

freakboy3742 commented on Misc/NEWS:
As noted previously - this file shouldn't exist in the patch.

johnzhou721 replied:
Will remove this after I get home and make another commit

johnzhou721 replied:
Deleted!

freakboy3742 commented on config.sub:
```diff
@@ -1743,7 +1743,7 @@ case $os in
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* | tvos* | watchos* \
+	     | os9* | macos* | osx* | ios* | tvos* | watchos* | xros* | visionos* \
```
I'm confused why both `xros` and `visionos` are listed here. The clang triple will always be `xros`.

johnzhou721 replied:
We’re lowercasing the OS name to get this part should we add a check just for xros?

freakboy3742 replied:
I'm unclear what you're asking - but this check should only include `xros`. The checks for tvOS don't check for `appletvos` in addition to `tvos`.

johnzhou721 replied:
I know, but in the Makefile we pass the relevant part of the platform triplet in as the OS name capitalized in PAS repo

johnzhou721 replied:
@freakboy3742 so the triplet is xrOS not visionOS? I mean it compiles with visionOS as well

freakboy3742 replied:
Ok - so we need to fix that. The same problem exists in the source deps repo, so it's going to need to be fixed there, too.

freakboy3742 replied:
> @freakboy3742 so the triplet is xrOS not visionOS? I mean it compiles with visionOS as well

Every source I can find online is using `arm64-apple-xros`, and all the SDK setting suggest that the preferred name is `xros`. If `arm64-apple-visionos` also works that's an annoying inconsistency - but I don't think we need to support it (or, if we do, we should be normalising the `visionos` value to `xros`, in the same way that `arm64` is normalised to `aarch64` for... GNU reasons).

johnzhou721 commented:
Oh sorry! Yeah I realized 5 seconds after I sent the ping comment that you were from australia

I was from China UTC+8, moved a few years (<5) ago, and yes, time zones are relatable

johnzhou721 commented:
Something along these lines: visionOS 3D content _must_ be done with SwiftUI and UIKit cannot be used for visionOS-specific or 3D content: https://developer.apple.com/documentation/visionos/bringing-your-app-to-visionos/

So... should we start writing rubicon-swift?

freakboy3742 commented on visionOS/Resources/bin/arm64-apple-visionos-clang:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+xcrun --sdk xros${XROS_SDK_VERSION} clang -target arm64-apple-visionos "$@"
```
Apple's tools accept `xros`, not `visionos`:
```
xcrun --sdk xros${XROS_SDK_VERSION} clang -target arm64-apple-xros "$@"
```
Similar changes are needed for the other tools.

johnzhou721 replied:
But I think visionOS is accepted on my latest version of Xcode, else clang will throw an error about it, which it doesn’t when I compile a small hello world c file with this target triplet.

johnzhou721 replied:
Nevertheless it would not hurt to have xrOS for consistency

freakboy3742 replied:
As noted in the comments about min-OS version handling, this will also need to say `-target arm64-apple-xros${XROS_DEPLOYMENT_VERSION}`.

johnzhou721 replied:
@freakboy3742 Do you mena XROS_DEPLOYMENT_TARGET not XROS_DEPLOYMENT_VERSION?

freakboy3742 commented on visionOS/Resources/dylib-Info-template.plist:
```diff
@@ -0,0 +1,26 @@
+		<string>xrOS</string>
+	</array>
+	<key>MinimumOSVersion</key>
+	<string>4.0</string>
```
Again, v4.0 is a problem.

That said - it also suggests a bigger problem, as this value should be dynamically generated from `XROS_DEPLOYMENT_VERSION`.

johnzhou721 replied:
Sorry! I copy and pasted stuff from watchOS, but now I feel guilty

freakboy3742 replied:
Ok - but then you actually need to fix it, not just mark the conversation as resolved :-)

freakboy3742 commented on configure.ac:
```diff
@@ -336,6 +336,9 @@ then
 	*-apple-watchos*)
 		ac_sys_system=watchOS
 		;;
+	*-apple-xros*|*apple-visionos*)
```
There's only one canonical form for the input triple - we should only be accepting the correct version.

johnzhou721 replied:
So xrOS?

freakboy3742 replied:
no - `*-apple-xros`.

johnzhou721 replied:
Yeah that was what I meant sorry for the confusion but since there’s only xr an vision

freakboy3742 commented on configure.ac:
```diff
@@ -906,6 +950,26 @@ if test "$cross_compiling" = yes; then
+		_host_device=`echo $host | cut -d '-' -f4`
+		_host_device=${_host_device:=os}
+
+		# VISIONOS_DEPLOYMENT_TARGET is the minimum supported visionOS version
```
As before - `XROS_DEPLOYMENT_TARGET` is the variable that Apple honours.

johnzhou721 replied:
Will fix when i get home late this evening (probably a bit before noon for you)

freakboy3742 commented on configure.ac:
```diff
@@ -5298,16 +5370,17 @@ if test "$MACHDEP" != linux; then
-# tvOS/watchOS have some additional methods that can be found, but not used.
-if test "$ac_sys_system" != "tvOS" -a "$ac_sys_system" != "watchOS" ; then
+# tvOS/watchOS/visionOS have some additional methods that can be found, but not used.
+dnl TODO: Can visionOS use those iOS only functions???
```
The test for this (and other checks in this general area) is whether the code will compile. If you can remove visionOS from this list, and the code still compiles, then the symbols are likely OK. They might need to be added back in if you run a test suite and they fail at runtime.

johnzhou721 replied:
Testing for this specific block right now.

johnzhou721 replied:
Compiled this block without visionOS on simulator already; testing device

johnzhou721 replied:
oopsies forgot to autoreconf

johnzhou721 replied:
ok Device is compiling fine, sorry for the delay, forgot to run make clean before reconfiguring and make almost got done but then got an error. Now for simulator

johnzhou721 replied:
finished building; committing,

johnzhou721 replied:
Testing the other tty block.

I have a question -- will visionOS ever be more general than iOS? Like will it have more commands than iOS? I just tested the ones that are disabled on tvOS and watchOS only.

johnzhou721 replied:
Done.

freakboy3742 commented on configure.ac:
```diff
@@ -1098,7 +1165,8 @@ AS_CASE([$host],
 
-dnl Add the compiler flag for the iOS/tvOS/watchOS minimum supported OS version.
+dnl Add the compiler flag for the iOS/tvOS/watchOS/visionOS minimum supported OS version,
+dnl albeit Apple now encodes the visionOS value into the target triplet.
```
It can also be passed in as `-mtargetos=xros1.0` (see [this change to Rust-cc](https://github.com/rust-lang/cc-rs/pull/1339/files#diff-7b19987f13f75d72d1f447086be16710eb6330767898f3493bf54ac2b15fa0c1R697))

johnzhou721 replied:
Apparently my google skills of those things are Rusting away…

freakboy3742 replied:
That said - it looks like specifying `-target arm64-apple-xros` *and* `-mtargetos=xros1.0` is not permitted... there's a [followup commit in rustcc](https://github.com/rust-lang/cc-rs/pull/1384) that clarifies the argument usage.

So - we either need to use `-target arm64-apple-xros2.0` *or* `-arch arm64 -mtargetos=xros2.0`. 

Given the way that we need to use these flags, and the usage in the compiler shim, my inclination is to use `-target arm64-apple-xros${XROS_DEPLOYMENT_VERSION}`

johnzhou721 replied:
So is leaving it like this okay? Or am I not experienced enough to add comments?

johnzhou721 replied:
Resolving for now.

freakboy3742 reviewed (changes requested):
I've flagged a bunch of consistency issues - mostly related to where the build scripts should be accepting `visionOS`, and when they should be accepting `xros`. However, there's also a bunch of 4.0 version references that I flagged in my last pass.

freakboy3742 commented:
> Something along these lines: visionOS 3D content _must_ be done with SwiftUI and UIKit cannot be used for visionOS-specific or 3D content: https://developer.apple.com/documentation/visionos/bringing-your-app-to-visionos/
> 
> So... should we start writing rubicon-swift?

See beeware/rubicon-objc#563.

The same problem/limitation exists for watchOS.

johnzhou721 commented:
Just finished changing a bunch of stuff; sorry for the delay; will push once build finishes
Also thanks for doing some final touchups and merging the cpython-apple-source-deps stuff.

johnzhou721 commented:
Also I got this while compiling:

./Modules/_hacl/Lib_Memzero0.c:52:6: warning: "Your platform does not support any safe implementation of memzero -- consider a pull request!" [-W#warnings]
   52 |     #warning "Your platform does not support any safe implementation of memzero -- consider a pull request!"
      |      ^

johnzhou721 committed:
xrOS

johnzhou721 committed:
remove unnessacary disablement of commands

johnzhou721 committed:
more enablement

johnzhou721 commented:
I have a question -- will visionOS ever be more general than iOS in terms of available commands? Like will it have more commands than iOS? I just tested the ones that are disabled on tvOS and watchOS only.

Anyway this is ready for re-review.

johnzhou721 committed:
Add platform triplet

freakboy3742 commented on configure.ac:
```diff
@@ -1358,6 +1415,9 @@ AS_CASE([$host/$ac_cv_cc_name],
   [aarch64-apple-tvos*/clang],            [PY_SUPPORT_TIER=3], dnl tvOS on ARM64
   [aarch64-apple-watchos*-simulator/clang],   [PY_SUPPORT_TIER=3], dnl watchOS Simulator on arm64
   [arm64_32-apple-watchos*/clang],            [PY_SUPPORT_TIER=3], dnl watchOS on ARM64
+dnl not adding the xros triplet, even though it's recognized
```
I'm not sure I understand what you're saying with this comment

johnzhou721 replied:
Oh, leftover from last time when I did visionos instead of xros

johnzhou721 commented:
Given that this builds, I am inclined to mark it as undraft.

freakboy3742 commented on configure.ac:
```diff
@@ -1098,7 +1153,8 @@ AS_CASE([$host],
 )
 
-dnl Add the compiler flag for the iOS/tvOS/watchOS minimum supported OS version.
+dnl Add the compiler flag for the iOS/tvOS/watchOS/visionOS minimum supported OS version,
+dnl albeit Apple now encodes the visionOS value into the target triplet.
```
```suggestion
dnl Add the compiler flag for the iOS/tvOS/watchOS minimum supported OS 
dnl version. visionOS doesn't use an explicit -mxros-version-min option - 
dnl it encodes the min version into the target triple.
```

johnzhou721 replied:
Great

freakboy3742 reviewed (changes requested):
A couple of minor clarifications, but this all looks like it's heading in the right direction. 

Procedurally, I won't accept and merge until I can verify the patch actually works in the context of the Python-Apple-support repo.

johnzhou721 commented:
Regarding dylib-info-template: I thought, wait what iisn't it already like this in _Info_.Plist.in? Sorry for the file confusion.

johnzhou721 committed:
comments and version

johnzhou721 commented:
Done.

johnzhou721 commented:
I put 2.0 there because I don't know how to setup a .in file or something like that.

freakboy3742 commented:
> Regarding dylib-info-template: I thought, wait what iisn't it already like this in _Info_.Plist.in? Sorry for the file confusion.

The two files are very similar, but have a different purpose. `Info.plist` is the plist file for the `Python.framework` that is generated; `dylib-info-template.plist` is, as the name suggests, a template for the plist that is used for each framework that is extracted as part of the build process. 

> I put 2.0 there because I don't know how to setup a .in file or something like that.

Sure - that's fine for now. There's clearly a deeper problem that needs to be fixed, as the same issue exists with the iOS version of the file.

johnzhou721 commented:
And tvOS and watchOS :)
seriously all computers smartwatches phones should use the same ABI so we wouldn't have to waste our breaths here doing trivial things. Not that I do not want to do them; I am really happy to help and am so glad I did

johnzhou721 commented:
btw I doubt if this is a coincidence, but Python str(round(pi, 2)) + "a6" released on pi day (March 14)...

Coincidence?

freakboy3742 commented:
> btw I doubt if this is a coincidence, but Python str(round(pi, 2)) + "a6" released on pi day (March 14)...
> 
> Coincidence?

No, it was not a coincidence :-)

johnzhou721 commented:
Anyway test is failing on testbed modified for a modified iOS testbed for visionos after initializing Python runtime it errors, I printed the full error over Python C API
Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 23, in <module>
TypeError: cannot unpack non-iterable function object

But importlib._bootstrap line 23 is just a function def

https://github.com/freakboy3742/cpython/blob/main/Lib/importlib/_bootstrap.py

It's also the first non-empty line which is suspicious

Any more suggestions before we deal with PAS repo stuff?

freakboy3742 commented:
> Any more suggestions before we deal with PAS repo stuff?

No particular suggestions; however, if you've got a modified testbed that can be used to prove that this set of changes works, then it would be good to include that testbed in the patch. Ideally, we'd have the same set of testbeds for tvOS and watchOS as well.

Regarding the specific error you're seeing - I can't say I have any particular insight there, other than turn up verbosity in the test scripts and see what else it reports.

johnzhou721 committed:
export visionOS deployment

johnzhou721 committed:
add (probably) nonworking testbed

johnzhou721 commented:
It proves that this set of changes does _not_ work, actually.

freakboy3742 reviewed (changes requested):
1. The configure script fails if you don't explicitly pass in ``--enable-ipv6``. There's a test on line 5852 of the configure script that checks for `ac_cv_buggy_getaddrinfo`; it currently checks for iOS; that should be extended for visionOS (and probably for tvOS, watchOS as well - that's an oversight of the existing patch)
2. If we're adding a testbed, that testbed should be named `visionOSTestbed`, not `iOSTestbed`
3. If we've got a testbed, there should also be a `testvisionos` Makefile target.
4. The visionOS testbed still has a lot of references to iOS in it

Longer term, there's definitely a need to consolidate the testbed script so that identical code isn't duplicated between 4 different platforms - but that's not a problem we need to solve here.

As for the failure at runtime: my guess is that it's related to L1538 of `_bootstrap_external.py` - visionOS isn't on the list of platforms that is being identified as needing an AppleFrameworkLoader, so it's not able to complete loading the bootstrap. That likely indicates a lot of other places where `visionos` is missing from platform checks.

johnzhou721 commented:
Yes, and also somehow this is building .so files. Will debug

johnzhou721 commented:
@freakboy3742 I mirrored the places in the Python.patch file for places that has ``watchOS`` or ``watchos`` and adds ``visionOS`` or ``xros``  there -- probably missed a few of these I guess.

freakboy3742 commented:
> Yes, and also somehow this is building .so files. Will debug

That's entirely expected. iOS produces .so as well.

johnzhou721 committed:
consistency

johnzhou721 commented:
Addressed 1, 2, 4 in the last commit. there's probably more 4 stuff will leave 3 for tmororow

johnzhou721 commented:
Hi! Sorry, had some busy schoolwork and didn't work as expected.

Command I'm building with (so I don't have to dig through my .zsh_history file again, just for my reference) without using BZip2 et al:
``export PATH=./visionOS/Resources/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin``
``./configure --host=arm64-apple-xros2.0-simulator --build=arm64-apple-darwin --enable-framework --with-build-python=/Library/Frameworks/Python.framework/Versions/3.14/bin/python3.14 && make && make install``

johnzhou721 commented:
And for device:
```
export PATH=./visionOS/Resources/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
./configure --host=arm64-apple-xros2.0--build=arm64-apple-darwin --enable-framework --with-build-python=/Library/Frameworks/Python.framework/Versions/3.14/bin/python3.14 && make && make install
```

Those are just for my own reference.

johnzhou721 committed:
Add testvisionos

johnzhou721 committed:
Remove more iOS refs.

johnzhou721 commented:
@freakboy3742 I just mirroed the watchOS and tvOS patch for making adjustments to Lib/ but apparently you had more when contributing to core cpython... Will go look

johnzhou721 committed:
Platform identification on visionOS.

johnzhou721 commented:
Just went through https://github.com/python/cpython/pull/117052 and https://github.com/python/cpython/pull/116454 and went ahead and searched for ``watchOS`` and added ``visionOS`` or ``xrOS`` where appropriate. I also spotted a place where I think you forgot to add watchos and tvos deployment targets to a list of stuff that should not be converted to integers in sysconfig; went ahead and fixed that in 4cd23bd.

johnzhou721 commented:
Tests, however, are still reporting the same error and not starting up. Did I forget to run some step I would have to do after I change importlib?

freakboy3742 commented:
There is a makefile target (`make regen-importlib`) that you can run; however, as of Python 3.12 (or thereabouts), it's automatically run as part of a build.

As for the patch being complete - A really cursory search for `tvos` and `watchos` in the source reveals a bunch of places where visionOS hasn't been added. The most glaring is `Lib/subprocess.py`; there's also a lot of them in `ctypes`; but they're not the only ones. I don't know if that's the contributing cause, but it's at least somewhere to start looking.

If that doesn't fix the startup issue, I'll need to go digging. Unfortunately, debugging the frozen libraries is complicated under the best of circumstances, and simulated Apple hardware isn't the best of circumstances :-)

freakboy3742 commented:
Oh - and the testbed is still referencing `xros-arm64_x86_64-simulator`, which is clearly wrong as well.

johnzhou721 commented:
Will fix tomorrow in your time zone b/c will be busy today. Happy early easter by the way

johnzhou721 commented:
I mean I will probably still be busy in the evening, and since we're talking about *frozen* modules right now, I might have to *Let it Go*.

johnzhou721 committed:
Add more visionos references that were missing.

johnzhou721 commented:
Anyway just pushed a commit addressing more refs. Hoping this will go well

johnzhou721 committed:
fix an outdated re

johnzhou721 committed:
Remove some debug statements

johnzhou721 commented:
Just removed some stuff for debuggin I shouldn't have committed... this is still failing with this on a fresh checkout of main (after `git clean -dfx`) using the command below -- I didn't do the BZip2 etc stuff yet. Let's update patch after we find out what's going on here.

```
export PATH=./visionOS/Resources/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
./configure --host=arm64-apple-xros2.0-simulator --build=arm64-apple-darwin --enable-framework --with-build-python=/Library/Frameworks/Python.framework/Versions/3.14/bin/python3.14 && make -j16 && make install && make testvisionos
```

johnzhou721 commented:
The interesting thing is that this is failing on iOS as well.

freakboy3742 commented:
> The interesting thing is that this is failing on iOS as well.

Well that's an indication that there's *something* in this patch that is broken - because the unpatched version is being executed and is passing on every commit to CI, and the builds from the `3.14-patched` branch are passing CI on the Python-Apple-support branch.

johnzhou721 commented:
But... What?

On Sat, Apr 19, 2025 at 1:20 AM Russell Keith-Magee <
***@***.***> wrote:

> The interesting thing is that this is failing on iOS as well.
>
> Well that's an indication that there's *something* in this patch that is
> broken - because the unpatched version is being executed and is passing on
> every commit to CI, and the builds from the 3.14-patched branch are
> passing CI on the Python-Apple-support branch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freakboy3742/cpython/pull/6#issuecomment-2816565602>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BRODAI5VSMRQRMK35WII3SD22HTMLAVCNFSM6AAAAAB3AVCZW2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWGU3DKNRQGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> *freakboy3742* left a comment (freakboy3742/cpython#6)
> <https://github.com/freakboy3742/cpython/pull/6#issuecomment-2816565602>
>
> The interesting thing is that this is failing on iOS as well.
>
> Well that's an indication that there's *something* in this patch that is
> broken - because the unpatched version is being executed and is passing on
> every commit to CI, and the builds from the 3.14-patched branch are
> passing CI on the Python-Apple-support branch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freakboy3742/cpython/pull/6#issuecomment-2816565602>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BRODAI5VSMRQRMK35WII3SD22HTMLAVCNFSM6AAAAAB3AVCZW2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWGU3DKNRQGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

johnzhou721 commented:
Hold on let me build with -O0 in cflags and then use the debugger

johnzhou721 committed:
lower/upper

johnzhou721 commented:
OK I still have absolutely no idea what the heck is going on\dots{} could you please take a look? I'm sorry but my familiarity with the CPython source is pretty limited -- I know how the autoconf system works though (even though I probably don't know a lot of the huge pile of macros .ac files uses) -- could you please take a look? Or am I simply not experienced enough to keep hanging in here?

johnzhou721 commented on configure.ac:
```diff
@@ -831,6 +863,8 @@ AC_ARG_WITH(
 ])
 AC_SUBST([APP_STORE_COMPLIANCE_PATCH])
 
+EXPORT_XROS_DEPLOYMENT_TARGET='#'
```
Github highlights the # as commenting out the rest of the line -- does this actually happen? My understanding is that within single quotes those stuff are treated verbatim.

johnzhou721 commented on configure.ac:
```diff
@@ -5298,16 +5366,16 @@ if test "$MACHDEP" != linux; then
 
 # tvOS/watchOS have some additional methods that can be found, but not used.
-if test "$ac_sys_system" != "tvOS" -a "$ac_sys_system" != "watchOS" ; then
+if test "$ac_sys_system" != "tvOS" -a "$ac_sys_system" != "watchOS"; then
```
Does the removal of the space matter? I accidentally did it after like reverting adding visionos

johnzhou721 commented on configure.ac:
```diff
@@ -5566,7 +5634,8 @@ PY_CHECK_FUNC([setgroups], [
 # tvOS/watchOS have functions for tty, but can't use them
-if test "$ac_sys_system" != "tvOS" -a "$ac_sys_system" != "watchOS" ; then
+dnl is the same applies for visionOS
+if test "$ac_sys_system" != "tvOS" -a "$ac_sys_system" != "watchOS"; then
```
Does the removal of the space matter? I accidentally did it after like reverting adding visionos
Also need to remove dnl above

johnzhou721 commented on configure.ac:
```diff
@@ -7033,7 +7102,7 @@ AC_MSG_NOTICE([checking for device files])
 dnl Some cross-compile builds are predictable; they won't ever
 dnl have /dev/ptmx or /dev/ptc, so we can set them explicitly.
-if test "$ac_sys_system" = "Linux-android" -o "$ac_sys_system" = "iOS" -o "$ac_sys_system" = "tvOS" -o "$ac_sys_system" = "watchOS" ; then
+if test "$ac_sys_system" = "Linux-android" -o "$ac_sys_system" = "iOS" -o "$ac_sys_system" = "tvOS" -o "$ac_sys_system" = "watchOS"  -o "$ac_sys_system" = "visionOS"; then
```
Same thing -- is space needed before ;?

johnzhou721 commented:
Flagged a couple of sus things that might impact logic. Does any of those matter?

johnzhou721 commented:
Nope. Nevermind -- I didn't really research about bash syntax before opening those useless stuff.

johnzhou721 committed:
code styl

johnzhou721 commented:
Still I pushed a commit to address the code style.

freakboy3742 commented:
I've just tried the laster version of this branch - and it's launching fine for me, on both iOS and visionOS, under Xcode.

It's possible the cause of the problem you're seeing is that you need a "build" version of Python that must be the *exact* same version that is being used for testing. So - you need to build a macOS version of CPython off the patched branch, and that needs to be the version of CPython that the visionOS build finds - essentially your "build" python needs to be on the path before the version in `/Library/Frameworks/Python.framework`. If this isn't the case, the frozen modules may not be compatible, which would lead to the sort of error you're reporting on startup.

If you're running the test suite through Xcode, you'll also need to [turn on automated breakpoints](https://github.com/freakboy3742/cpython/blob/main/iOS/README.rst#disabling-automated-breakpoints), same as you do for iOS; without that, the test suite will halt about 5 tests in when then first SIGINT is raised

With those two caveats, most of the test suite is running; however:
* When the test suite is building, I'm seeing a warning about compiling InterfaceBuilder products. I suspect this is occurring because you've taken the iOS test project verbatim; some cleanup is needed here.
* When the test suite runs, I'm seeing at least 4 test failures due to basic platform identification issues (issues where a platform tag is coming back as `visionos-2.0-visionsimulator` not `visionos-2.0-xrsimulator`, or similar).
* I'm seeing a hard crash near the end of the test suite. I haven't dug into detail to find the cause of this one.
* The `run` target on a testbed project (i.e., `path/to/testbed run`), and by extension, `make testvisionos`, isn't working for me. It's timing out after 5 minutes; it looks to me like it is *starting* a simulator, but the mechanism in the testbed that tries to identify the booted simulator isn't working.

johnzhou721 committed:
Remove reference to LaunchScreen storyboard.

johnzhou721 committed:
fix simulator detection

johnzhou721 commented:
1. Can confirm stuff related to using the official alpha 7 build gives the error.

2. What are we supposed to do with this thing? Just continued for now but it says ``Message from debugger: Terminated due to signal 9``. Is this the hard crash you are talking about?
<img width="1840" alt="Screenshot 2025-04-21 at 9 42 10 AM" src="https://github.com/user-attachments/assets/5df30470-63e9-49f3-acc0-5e4742a41a8f" />

3. You said "needs to be on the path before the version in ``/Library/Frameworks/Python.framework`` ", but it wasn't on the path at all, I just passed it to ``--with-build-python`` (wasn't using Python-Apple-Support), is that fine?

4. Just pushed a commit addressing item 1.

5. 4 seems to be caused by the devices being listed using xrOS and not visionOS in simctl, fixed.

I'm testing this right now... I hope messages like this counts as a small amount of large messages.

johnzhou721 commented:
Caught where the issue for #2 is

Current commands for build --

```
export PATH=./visionOS/Resources/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
./configure --host=arm64-apple-xros2.0-simulator --build=arm64-apple-darwin --enable-framework --with-build-python=/Users/johnzhou/cpythonxrmac/bin/python3.14 && make -j16 && make install -j16 && make testvisionos
```

johnzhou721 committed:
fix some more tests

freakboy3742 commented:
> 1. Can confirm stuff related to using the official alpha 7 build gives the error.

That would be consistent. Any new feature that alters CPython's bytecode definition will be a source of problems, and a7 included the addition of t-strings. Once beta 1 lands, these problems become less prevalent because new features are no longer allowed.

> 2. What are we supposed to do with this thing? Just continued for now but it says `Message from debugger: Terminated due to signal 9`. Is this the hard crash you are talking about?

Yes, that's the hard crash I was talking about. 

As for how to handle it... it depends what is causing it. [`EXC_GUARD` is an exception that indicates access of a protected file system resource](https://developer.apple.com/documentation/xcode/exc_guard), which seems highly probable given the test that seems to be failing is in `test_zipfile`. The question is which resource is causing the problem.

So - the next step is to narrow down on which subtest of `test_zipfile` is the specific cause of the failure. If you've got the project loaded in Xcode, you can edit the `visionOSTestbed-Info.plist` file; that file contains a `TestArgs` key which are the arguments that will be passed to the test suite. Add `test_zipfile` `-v` as 2 additional arguments, and *only* `test_zipfile` will run, and it will run in "verbose" mode, showing each test as it runs. That should help you to narrow down to a specific test that is causing the error. Then you can dig into the code for [`test_zipfile`](https://github.com/freakboy3742/cpython/tree/main/Lib/test/test_zipfile) for that test, and work out what is going on. 

Beyond that, it depends what the test is doing to cause a problem. It might be as simple as a test that is trying to access a file that isn't available in the visionOS sandbox. At the very least, we can skip that specific test. However, ideally we won't just skip the test, we'd disable the fix the underlying problem. If the issue is the specific filename, then we can find a better filename; if it's using an underlying feature, we can disable that feature. That's what [these checks](https://github.com/freakboy3742/cpython/blob/main/configure.ac#L5185) are controlling - it's identifying system calls that *exist*, but if you actually try to invoke them, an error is raised. 
 
If you want to test a theory about whether disabling a specific test will fix the problem, you can manually edit the contents of `test_zipfile` in the package Framework, then do a clean build of the app - that will run your patched code in the test suite.

> 3. You said "needs to be on the path before the version in `/Library/Frameworks/Python.framework` ", but it wasn't on the path at all, I just passed it to `--with-build-python` (wasn't using Python-Apple-Support), is that fine?

Yes, that's fine - `--with-build-python` is the other way (and more reliably way) to guarantee the interpreter that is used.

> 5. 4 seems to be caused by the devices being listed using xrOS and not visionOS in simctl, fixed.

That's about what I expected; glad to see it was an easy one.
 
> I'm testing this right now... I hope messages like this counts as a small amount of large messages.

Yes, this is a lot better - thanks for taking that feedback on board.

freakboy3742 commented on visionOS/testbed/visionOSTestbed/dylib-Info-template.plist:
```diff
@@ -0,0 +1,26 @@
+	<key>CFBundleShortVersionString</key>
+	<string>1.0</string>
+	<key>CFBundleSupportedPlatforms</key>
+	<array>
+		<string>iPhoneOS</string>
+	</array>
+	<key>MinimumOSVersion</key>
+	<string>12.0</string>
```
Just noticed that these keys don't seem right

johnzhou721 replied:
Not sure about canoncial input for this, and don't know how to look it up (Google is failing me), but by searching up random stuff I managed to find that the tvOS value is AppleTVOS, so I used VisionOS as of right now.

johnzhou721 commented:
Also it's funny ``#2`` was referencing issue ``2``. Anyway I just pushed a change to Python-Apple-Support doing a macOS build in the CI, so we could test these stuff on CI before Beta 1 comes out. If we happen to merge after (or before and you don't like it, I'm guessing so) Beta 1, we will revert.

johnzhou721 committed:
plist cleanup

johnzhou721 commented:
Let's go iOS passing for the first time in forever on PAS...

Will go investigate issue about zipfile now. Doing what you said; however, FYI your comment has a formatting issue, but don't worry i've had tons during translation.

johnzhou721 commented:
@freakboy3742 Sorry for sending a premature message 14 minutes ago.

I set verbose + test_zipfile but it crashes without printing anything verbose (aka just the test name). Any ideas?
<img width="1840" alt="Screenshot 2025-04-21 at 6 50 39 PM" src="https://github.com/user-attachments/assets/e8055a8f-71d1-4fdf-bc95-38adc5a7f98a" />
<img width="1840" alt="Screenshot 2025-04-21 at 6 52 18 PM" src="https://github.com/user-attachments/assets/c50011fa-b313-4e37-9fbf-14a2ee52e0f8" />
Any ideas of what to do now?

Also never mind test_platform is failing, a reference to the fact that I was writing in C++ 5 minutes ago.

johnzhou721 committed:
fix typo

freakboy3742 commented:
> Any ideas of what to do now?

My apologies - I forgot that `-W` and `-v` interact. You'll also need to delete the `-W` option. You should get output that looks like:

```
...
test_write (test.test_zipfile.test_core.UnseekableTests.test_write) ... ok
test_writestr (test.test_zipfile.test_core.UnseekableTests.test_writestr) ... ok
test_compresslevel_property (test.test_zipfile.test_core.ZipInfoTests.test_compresslevel_property) ... ok
test_from_dir (test.test_zipfile.test_core.ZipInfoTests.test_from_dir) ... ok
test_from_file (test.test_zipfile.test_core.ZipInfoTests.test_from_file) ... ok
test_from_file_bytes (test.test_zipfile.test_core.ZipInfoTests.test_from_file_bytes) ... ok
test_from_file_fileno (test.test_zipfile.test_core.ZipInfoTests.test_from_file_fileno) ... ok
test_from_file_pathlike (test.test_zipfile.test_core.ZipInfoTests.test_from_file_pathlike) ... ok

----------------------------------------------------------------------
Ran 333 tests in 7.470s

OK (skipped=3)

== Tests result: SUCCESS ==

1 test OK.

Total duration: 7.6 sec
Total tests: run=333 skipped=3
Total test files: run=1/1
Result: SUCCESS
```

johnzhou721 commented:
Hi @freakboy3742, apologies for the slow response time (and the many other times where I had to wake up before I could respond, even though I wake up as early as 5:30 in some days within the past month or so).

The crashing test is ``test.test_zipfile.test_core.TestsWithMultipleOpens.test_many_opens``, whose job is to verify
```
        # Verify that read() and open() promptly close the file descriptor,
        # and don't rely on the garbage collector to free resources.
```
Now, the code is pretty short that I can post it here (if GitHub ever relicenses user-submitted contents under dirty stuff we can argue fair use here).
```
    def test_many_opens(self):
        # Verify that read() and open() promptly close the file descriptor,
        # and don't rely on the garbage collector to free resources.
        startcount = fd_count()
        self.make_test_archive(TESTFN2)
        with zipfile.ZipFile(TESTFN2, mode="r") as zipf:
            for x in range(100):
                zipf.read('ones')
                with zipf.open('ones') as zopen1:
                    pass
        self.assertEqual(startcount, fd_count())
```
``fd_count`` seems sussy... any insights other than "disable the test"?

freakboy3742 commented:
> `fd_count` seems sussy... any insights other than "disable the test"?

I guess my follow up question is to ask what line is actually throwing the error? Is it `fd_count()`? That seems unlikely, as `fd_count()` is a test utility method used elsewhere in the test suite. Or is it one of the other calls? In particular - is `make_test_archive()` trying to create a file somewhere it's not allowed to? Is it rejecting the request to read the specific zipfile entry?

johnzhou721 commented:
By monkey patching (changing the produced XCFramework) and putting ``print("=======")`` (a technique from doing competitive programming (not-that-successfully)), I found that the culprit is ``fd_count`` on the line following where it's inserted
```
    def test_many_opens(self):
        # Verify that read() and open() promptly close the file descriptor,
        # and don't rely on the garbage collector to free resources.
        print("=======")
        startcount = fd_count()
        self.make_test_archive(TESTFN2)
        with zipfile.ZipFile(TESTFN2, mode="r") as zipf:
            for x in range(100):
                zipf.read('ones')
                with zipf.open('ones') as zopen1:
                    pass
        self.assertEqual(startcount, fd_count())
```
Will stop working today at GMT-5 (daylight saving time already applied into the time zone), and sorry for the brisk comment.

freakboy3742 commented:
> By monkey patching (changing the produced XCFramework) and putting `print("=======")` (a technique from doing competitive programming (not-that-successfully)), I found that the culprit is `fd_count` on the line following where it's inserted

Looks like I spoke too soon about `fd_count()` not being a likely culprit :-)

I've done a poke around, and all the exiting uses are being skipped - including, in one case, [this skip](https://github.com/freakboy3742/cpython/blob/3.14-patched/Lib/test/test_support.py#L551) that identifies the test as "unstable". 

I'd like to have a little better understanding of exactly why it's failing on visionOS, so if you're able to dig into the internals of `fd_count()` to see where it's failing that might point at a cause; but I'm getting close to just saying "skip it".

johnzhou721 commented:
> > By monkey patching (changing the produced XCFramework) and putting `print("=======")` (a technique from doing competitive programming (not-that-successfully)), I found that the culprit is `fd_count` on the line following where it's inserted
> 
> Looks like I spoke too soon about `fd_count()` not being a likely culprit :-)

Just went ahead and re-checked this using flush=True on print and yes I can confirm :)

> I've done a poke around, and all the exiting uses are being skipped - including, in one case, [this skip](https://github.com/freakboy3742/cpython/blob/3.14-patched/Lib/test/test_support.py#L551) that identifies the test as "unstable".

Yes, but that being skipped because it's unstable is because it produces different output because of system stuff tempering w/ the file descriptor count (which I have no idea about), not because it crashes AFAIK (by looking at the comments).

> I'd like to have a little better understanding of exactly why it's failing on visionOS, so if you're able to dig into the internals of `fd_count()` to see where it's failing that might point at a cause; but I'm getting close to just saying "skip it".

Crashing, not failing. In my own opinion I would not skip it. After extensive debugging, it is crashing on https://github.com/johnzhou721/cpython/blob/379c3d965724046bffcde63199893fe79dfdded7/Lib/test/support/os_helper.py#L689 for the 8th time it executes (fd=7, but 0-indexing :) )

<details>
 <summary>If you got confused on my stuff above and would like to see concrete evidence</summary>

Relevant definition (see the lines ``#ADD``) I added:
```
def fd_count():
    """Count the number of open file descriptors.
    """
    if sys.platform.startswith(('linux', 'android', 'freebsd', 'emscripten')):
        fd_path = "/proc/self/fd"
    elif sys.platform == "darwin":
        fd_path = "/dev/fd"
    else:
        fd_path = None

    if fd_path is not None:
        try:
            names = os.listdir(fd_path)
            # Subtract one because listdir() internally opens a file
            # descriptor to list the content of the directory.
            return len(names) - 1
        except FileNotFoundError:
            pass

    MAXFD = 256
    if hasattr(os, 'sysconf'):
        try:
            MAXFD = os.sysconf("SC_OPEN_MAX")
        except OSError:
            pass

    old_modes = None
    if sys.platform == 'win32':
        # bpo-25306, bpo-31009: Call CrtSetReportMode() to not kill the process
        # on invalid file descriptor if Python is compiled in debug mode
        try:
            import msvcrt
            msvcrt.CrtSetReportMode
        except (AttributeError, ImportError):
            # no msvcrt or a release build
            pass
        else:
            old_modes = {}
            for report_type in (msvcrt.CRT_WARN,
                                msvcrt.CRT_ERROR,
                                msvcrt.CRT_ASSERT):
                old_modes[report_type] = msvcrt.CrtSetReportMode(report_type,
                                                                 0)

    try:
        count = 0
        for fd in range(MAXFD):
            try:
                # Prefer dup() over fstat(). fstat() can require input/output
                # whereas dup() doesn't.
                print("========", fd, flush=True)             # ADDED BY ME
                fd2 = os.dup(fd)             # 689 IN ORIGINAL
                print("++++++++", fd, flush=True)             # ADDED BY ME
            except OSError as e:
                if e.errno != errno.EBADF:
                    raise
            else:
                os.close(fd2)
                count += 1
    finally:
        if old_modes is not None:
            for report_type in (msvcrt.CRT_WARN,
                                msvcrt.CRT_ERROR,
                                msvcrt.CRT_ASSERT):
                msvcrt.CrtSetReportMode(report_type, old_modes[report_type])

    return count
```

Log just before it crashed:
<img width="1840" alt="Screenshot 2025-04-22 at 6 46 00 PM" src="https://github.com/user-attachments/assets/b7593ea5-694e-4825-9742-3c249953d75f" />
</details>

freakboy3742 commented:
> > I'd like to have a little better understanding of exactly why it's failing on visionOS, so if you're able to dig into the internals of `fd_count()` to see where it's failing that might point at a cause; but I'm getting close to just saying "skip it".
> 
> Crashing, not failing. In my own opinion I would not skip it. After extensive debugging, it is crashing on https://github.com/johnzhou721/cpython/blob/379c3d965724046bffcde63199893fe79dfdded7/Lib/test/support/os_helper.py#L689 for the 8th time it executes (fd=7, but 0-indexing :) )

My apologies - that was loose language on my part. I'm aware it's a crash, not a soft fail.

That testing definitely confirms it's `os.dup()` that is the problem; and from some poking around, it's not *entirely* surprising that [`EXC_GUARD` is a problem for some file descriptors](https://developer.apple.com/documentation/xcode/exc_guard?language=objc). Unfortunately, Apple doesn't seem to want to document how these file descriptors *should* be handled.

The good news, though, is that the fix is staring us in the face. [CPython macOS has already hit this problem](https://github.com/python/cpython/issues/109981), and applied a fix. At the top of `fd_count()` is a short-cut mechanism for getting a descriptor count on `darwin`... but iOS, tvOS, watchOS and visionOS are also Darwin kernels, so the same approach works there too.

So - if you modify the `if sys.platform == 'darwin'` check to be `if support.is_apple`, you should get a pass on the `test_zipimport` test. That fix should probably be applied upstream as well, as it will be an improvement for iOS as well - so if you want to earn your CPython contribution wings, submit that fix upstream (as well as making it on this PR) and tag me and I'll get it merged in.

johnzhou721 committed:
test: Resolve situation on iOS regarding fd_count.

macOS Sonoma seems to have issues with support.fd_count, at
gh-109981. However, after experimenting using macOS Sonoma
on an attempt to support visionOS unofficially through another
project, the simulator is hard-crashing, leading to the con-
clusion that this probably applies on all Apple platforms.
Hence this patch.

johnzhou721 commented:
@freakboy3742  Done.
https://github.com/python/cpython/pull/132823

freakboy3742 committed:
Final test cleanups for visionOS.

freakboy3742 reviewed (approved):
I've just pushed a couple of minor updates to your patch - but with those changes, I'm getting a clean test run! Thanks for your work on this patch.

johnzhou721 reviewed (commented):
sorry, finishing up a pending review from forever ago

johnzhou721 commented:
Like I flagged those style things and forgot to click ``finish your review''
